### PR TITLE
fix: assinging `type` for core flags

### PR DESCRIPTION
**Pull request by snitin315**
**What kind of change does this PR introduce?**
tests

**Did you add tests for your changes?**
yes
**If relevant, did you update the documentation?**
no
**Summary**
add test cases for `--module-parser-javascript-url`

**Does this PR introduce a breaking change?**
No

**Other information**
Related thread https://github.com/webpack/webpack-cli/pull/2438#discussion_r579130089

### CODE REVIEW

**Review by alexander-akait — COMMENTED** [review]
Can you show me `configs` of this option (https://github.com/webpack/webpack/blob/master/lib/cli.js#L181)?

**Comment by snitin315** [comment]
Here in the snapshot - https://github.com/webpack/webpack/blob/2a0536cf510768111a3a6dceeb14cb79b9f59273/test/__snapshots__/Cli.test.js.snap#L2681-L2697

**Comment by alexander-akait** [comment]
As you can see we have a bug on our side, we should support boolean values https://github.com/webpack/webpack/blob/2a0536cf510768111a3a6dceeb14cb79b9f59273/test/__snapshots__/Cli.test.js.snap#L2695

**Comment by alexander-akait** [comment]
We need migration on `configs` from `simpleType`

**Comment by snitin315** [comment]
Yes, I will do it.

File: packages/webpack-cli/lib/webpack-cli.js [hunk 1]
```diff
@@ -399,17 +399,41 @@ class WebpackCLI {
...
+                return String;
+            } else if (type === 'number') {
+                return Number;
+            } else if (type === 'boolean' || type === 'reset') {
+                return Boolean;
+            } else if (type === 'RegExp') {
+                return RegExp;
+            } else if (type === 'enum') {
+                return normalizeType(simpleType);
+            }
```
Comment (snitin315) [inline packages/webpack-cli/lib/webpack-cli.js:413]: Tackling this `enum` type is tricky. Open to suggestions for improvement.

File: packages/webpack-cli/lib/webpack-cli.js [hunk 2]
```diff
@@ -399,17 +399,41 @@ class WebpackCLI {
...
+                  if (meta.configs.length > 1) {
+                      meta.type = [];
+                      meta.configs.forEach((config) => {
+                          const type = normalizeType(config.type, meta.simpleType);
+                          if (!meta.type.includes(type)) {
+                              meta.type.push(type);
+                          }
+                          // 'cache' have type as enum but simpleType as boolean
+                          if (config.type === 'boolean' || type === Boolean) {
+                              meta.negative = true;
```
Comment (snitin315) [inline packages/webpack-cli/lib/webpack-cli.js:429]: https://github.com/webpack/webpack/blob/2a0536cf510768111a3a6dceeb14cb79b9f59273/test/__snapshots__/Cli.test.js.snap#L34-L58

**Review by alexander-akait — COMMENTED** [review]
We have more bugs :smile:

**Comment by snitin315** [comment]
It's more complex than it seems 😞.

File: packages/webpack-cli/lib/webpack-cli.js (hunk 1 above)
Comment (alexander-akait) [reply packages/webpack-cli/lib/webpack-cli.js:413]: We have more types https://github.com/webpack/webpack/blob/master/lib/cli.js#L42

Comment (alexander-akait) [reply packages/webpack-cli/lib/webpack-cli.js:413]: for enum I think you need to run `typeof` and collect all possible types

File: packages/webpack-cli/lib/webpack-cli.js [hunk 3]
```diff
@@ -399,17 +399,61 @@ class WebpackCLI {
...
+                    if (!typeofValues.includes(typeof value)) {
+                        typeofValues.push(typeof value);
+                    }
+                });
+                if (typeofValues.length === 1) {
+                    return normalizeType(typeofValues[0]);
+                } else {
+                    const types = typeofValues.map((type) => normalizeType(type));
+                    return types;
+                }
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:415]: Please avoid multiple loops, just return `['Number', 'String', 'Boolean']` here for `enum`

File: packages/webpack-cli/lib/webpack-cli.js [hunk 4]
```diff
@@ -399,17 +399,61 @@ class WebpackCLI {
...
+                      } else {
+                          meta.type = allowedTypes[0];
+                      }
                   } else {
-                      meta.type = Boolean;
-                      meta.negative = !flag.endsWith('-reset');
+                      meta.type = normalizeType(meta.simpleType);
+                      if (meta.type === Boolean && !flag.endsWith('-reset')) {
+                          meta.negative = true;
+                      }
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:456]: Some options can be only negative, i.e. `--amd` (doesn't exists), we have only `--no-amd`

File: packages/webpack-cli/lib/webpack-cli.js [hunk 5]
```diff
@@ -399,17 +399,61 @@ class WebpackCLI {
...
+                              if (!allowedTypes.includes(type)) {
+                                  allowedTypes.push(type);
+                              }
+                          }
+                      });
+                      if (allowedTypes.length > 1) {
+                          meta.type = allowedTypes;
+                      } else {
+                          meta.type = allowedTypes[0];
+                      }
```
Comment (alexander-akait) [inline packages/webpack-cli/lib/webpack-cli.js:451]: No need extra loop here too

**Comment by alexander-akait** [comment]
I can help to finish if you want

**Comment by snitin315** [comment]
> I can help to finish if you want

That would be awesome, feel free to push a commit.

**Comment by alexander-akait** [comment]
Solved locally, tomorrow I will send a PR

**Comment by alexander-akait** [comment]
WIP on tests

**Comment by alexander-akait** [comment]
Tests are hard, I need more time on this, WIP, hope finish it tomorrow

**Comment by alexander-akait** [comment]
We really need rewrite tests, unfortunately I do not have time to rewrite the tests now, in the near future I will start to create roadmap and include example of testing for options

**Comment by alexander-akait** [comment]
Added todo for macos tests, looks jest exit from tests and don't wait IO from node.js

**Comment by alexander-akait** [comment]
I hate cache, again the problem with cache...

**Pull request merged**